proposals/0111-linear-types.rst

```diff
@@ -297,40 +297,21 @@ This proposal adds two new syntactical constructs:
 
     \x %'One -> x
 
-  This modifies the syntax entry for pattern with signature annotation
-  as follows as follows
+  This modifies the syntax entry for lambda expressions
+  as follows
 
   ::
 
-    pat -> pat [PREFIX_PERCENT btype] [:: type]
+    lexp -> \ apat1 ... apatn [PREFIX_PERCENT btype] -> exp
```

Comment:
I find this change surprising. I thought the direction we were taking was to be able to specify, for each pattern in a lambda-expression to be an arg-type as we've been calling it (that is, everything that goes to the left of an arrow).

But here we need several lambda-expressions if we have several multiplicities to bind. Maybe it's the right thing to do? I suppose linearity annotations on nested patterns don't make a ton of sense. Also it precludes giving explicit multiplicities in equations. Usually the multiplicities come from the type signature, but this seems to run contrary to the variable-in-patterns story and the idea of not relying on the `forall p.` binders in type signature to bind names. So actually, linearity annotations on nested patterns do make sense: they let me bind multiplicity variables.

At the very least I'd like some motivation. Maybe a paragraph in the _Alternatives_ section (“a previous version of this proposal had <previous syntax> but <justification>” would help?

Reply:
My main motivation is so that modifiers always precede the thing they modify... and this new version seemed simpler, because I didn't think that a multiplicity annotation on a nested pattern made any sense. But you're right that a multiplicity annotation on nested patterns might indeed make sense.

What would you think of

```
pat -> [PREFIX_PERCENT btype] pat [:: type]
```

? This would mean we write ` \ (%1 x) -> x` or `\ (%1 x :: a) -> (x :: a)`. Not so different from Idris, actually.

Reply:
I've had assumed that the modifier is on the _arrow_ (since the modifier always precedes the modified thing). This is why `x %1 -> ...` made sense to me. In fact, I find a bit weird to allow other patterns, I would expect having to write the following for a linear function that matches:

```
\x %1 -> case x of ...
```

Reply:
After further thought there is a difficulty with using multiplicity modifier to bind variables in patterns: the current specification would not let me infer a modifier's type. If I wrote (in the notation you are proposing)

```haskell
\(%p x) -> u
```

Then GHC would go “you naughty, naughty person, I don't know what kind of modifier this is”. And it would be quite right too.

So maybe modifiers are not the right way to do multiplicity variable binding…

I'm honestly not quite sure what conclusion to draw from this.

Reply:
Is `p` in scope? Ah, no: I see you want to *bind* `p`. No dice. I say: we just don't allow this. Use the syntax from #155/#238 to do that.

Reply:
Ah yes. I got confused. I was thinking: I need multiplication annotation to bind existentially-bound multiplicity variables. But it's just not true: if they are existentially bound, then they are an argument of a constructor, and we can bind it at this site.

Nevermind this objection, then.

Then probably I like `\ (%1 x) -> x`. Maybe it makes more sense as part of the lambda syntax? Rather than nested in patterns. It's a bit fuzzy in my mind right now.

Reply:
My goal here is to have a consistent way to understand `%blah` in Haskell source code, and that is for the `%blah` to affect the next thing in the source. I don't have a strong sense for the best syntax in this spot, so happy to follow any suggestion of yours as long as the `%1` precedes something reasonably understood to be part of the lambda.

Reply:
I have difficulties paging all this back in active memory. But maybe it doesn't matter: the design principle is now laid out, and if we find the need to adjust in the future, we can, we don't have to think about everything right now. (though we have to admit that record fields are a bit inconsistent, ah well…)

My favoured syntax for this particular case is `\ (%1 x) ->`, it's probably sufficient. Maybe not, but probably. So I suggest we go for it.